### PR TITLE
refactor(core): re-implement useAtom with useReducer

### PR DESCRIPTION
**Pull request by dai-shi**
uMS won't land, #686 doesn't work so far.
I feel like simpler solution with useReducer would work better.
This simplifies a lot. Also, fixes weird issues in devtools.

### CODE REVIEW

**Comment by dai-shi** [comment]
This improves https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-rendering results:

v1.3.2

```
  jotai
    with useTransition
      ✓ test 1: updated properly with transition (3562 ms)
      ✓ test 2: no tearing with transition (122 ms)
      ✓ test 3: ability to interrupt render
      ✕ test 4: proper branching with transition (7543 ms)
    with intensive auto increment
      ✕ test 5: updated properly with auto increment (13214 ms)
      ✕ test 6: no tearing with auto increment (4 ms)
```

33a40b0

```
  jotai
    with useTransition
      ✓ test 1: updated properly with transition (3555 ms)
      ✓ test 2: no tearing with transition (119 ms)
      ✓ test 3: ability to interrupt render (1 ms)
      ✕ test 4: proper branching with transition (8503 ms)
    with intensive auto increment
      ✓ test 5: updated properly with auto increment (3233 ms)
      ✕ test 6: no tearing with auto increment (2 ms)
```

**Comment by dai-shi** [comment]
I would like to test if react devtools work before merging. (codesandbox ci is not working. 😓 )

File: src/devtools/useAtomsSnapshot.ts [hunk 1]
```diff
@@ -19,20 +16,25 @@ export function useAtomsSnapshot(scope?: Scope): AtomsSnapshot {
...
-  const [store, , devMutableSource, devSubscribe] = scopeContainer
+  const [store, devStore] = scopeContainer
 
-  const atoms = useMutableSource(
-    devMutableSource,
-    // FIXME HACK creating new reference to force re-render
-    useCallback((devContainer) => [...devContainer.atoms], []),
-    devSubscribe
-  )
+  const [atomsSnaphost, setAtomsSnaphost] = useState<AtomsSnapshot>(new Map())
```
Comment (Thisen) [inline src/devtools/useAtomsSnapshot.ts:21]:
```suggestion
  const [atomsSnapshot, setAtomsSnapshot] = useState<AtomsSnapshot>(new Map())
```

File: src/devtools/useAtomsSnapshot.ts [hunk 2]
```diff
@@ -19,20 +16,25 @@ export function useAtomsSnapshot(scope?: Scope): AtomsSnapshot {
...
+    const callback = () => {
+      const { atoms } = devStore
+      const atomToAtomValueTuples = atoms
+        .filter((atom) => !!store[DEV_GET_MOUNTED]?.(atom))
+        .map<[Atom<unknown>, unknown]>((atom) => {
+          const atomState =
+            store[DEV_GET_ATOM_STATE]?.(atom) ?? ({} as AtomState)
+          return [atom, atomState.v]
+        })
+      setAtomsSnaphost(new Map(atomToAtomValueTuples))
```
Comment (Thisen) [inline src/devtools/useAtomsSnapshot.ts:32]:
```suggestion
      setAtomsSnapshot(new Map(atomToAtomValueTuples))
```

File: src/devtools/useAtomsSnapshot.ts [hunk 3]
```diff
@@ -19,20 +16,25 @@ export function useAtomsSnapshot(scope?: Scope): AtomsSnapshot {
...
+  }, [store, devStore])
 
-  const atomToAtomValueTuples = atoms
-    .filter((atom) => !!store[DEV_GET_MOUNTED]?.(atom))
-    .map<[Atom<unknown>, unknown]>((atom) => {
-      const atomState = store[DEV_GET_ATOM_STATE]?.(atom) ?? ({} as AtomState)
-      return [atom, atomState.v]
-    })
-  return new Map(atomToAtomValueTuples)
+  return atomsSnaphost
```
Comment (Thisen) [inline src/devtools/useAtomsSnapshot.ts:39]:
```suggestion
  return atomsSnapshot
```

File: src/core/useAtom.ts [hunk 4]
```diff
@@ -75,8 +53,33 @@ export function useAtom<Value, Update>(
...
+    if (atomState.w) {
+      throw atomState.w // write promise
+    }
+    if ('v' in atomState) {
+      return atomState.v as Value
+    }
+    throw new Error('no atom value')
+  }, [store, atom])
+
+  const [value, forceUpdate] = useReducer(getAtomValue, undefined, getAtomValue)
```
Comment (Thisen) [inline src/core/useAtom.ts:75]: I never tried using a callback function as the reducer for `useReducer` - What happens when this callback is recreated due to dependency changes - Is it run again or only on dispatch?

File: src/devtools/useAtomsSnapshot.ts (hunk 1 above)
Comment (dai-shi) [reply src/devtools/useAtomsSnapshot.ts:21]: oops, nice catch.

File: src/core/useAtom.ts (hunk 4 above)
Comment (dai-shi) [reply src/core/useAtom.ts:75]: It doesn't have to be wrapped by `useCallback`. For more info, [this blog post](https://overreacted.io/a-complete-guide-to-useeffect/#why-usereducer-is-the-cheat-mode-of-hooks) might help.
I'm not saying this is ideal, but it's simpler than the previous uMS emulation and more readable. I'm sure we will revisit this in the future.

**Review by Thisen — APPROVED** [review]
Looks good to me. Surprised that it didn't require test changes.

**Pull request merged**